Vm to docker/images (#378)
netgroup-polito/CrownLabs: provisioning/containers/gui-common/README.md

# Docker-based GUI

> This sub-project aims to provide a Dockerized environment where Docker-based GUI applications can run. This is backed by a set of different containers, each with its own specific functions, that together constitute and offer the Docker-based GUI application.

Reviewer: This sentence looks ok, but not here. Probably is much better in the (parent) "containers" folder.

## General architecture

> In order to deploy GUI applications in CrownLabs the following is required:

Reviewer: "the following" what?
Maybe "he following components are required"?

> - A VNC-based X server

Reviewer: "A VNC-based X server"
or "An X11 server with VNC support"?

> - A window manager (it's better to have it since sub-windows or menu opened by the application might not get rendered correctly)

Reviewer: ```suggestion
- A window manager (required to allow sub-windows or menu opened by the application to be rendered correctly)
```

- Websockify
- noVNC client files (static HTML files hosted somewhere)
- the GUI application

> 

Reviewer: Maybe not here, but you have to document how those components are installed and run. Everything in the same Docker? Or in multiple Dockers? (e.g., sidecars)?

## Current implementation

Currently the following containers are used to build the overall GUI application:

- Xvnc server (based on TigerVNC) + fluxbox window manager, which are responsible to provide the basic graphical desktop environment.

> - Websockify (C version), exposing VNC server as a web server

Reviewer: Document (maybe not here, but it has to be said) why you're using the C version.

Technically, this consists in two different servers (the actual X server and a VNC server) that are bound together and provide a virtual desktop both to the user (through the VNC protocol) and to the final application (through the X core protocol).

There are several implementations of such type of server. For CrownLabs, [TigerVNC](https://github.com/TigerVNC/tigervnc) has been chosen, since it provides automatic desktop rescaling features.
Note that TigerVNC server cannot be considered a regular X server since it cannot display the desktop on a physical screen.

This container also includes [Fluxbox](http://fluxbox.org/), a lightweight window manager which provides an effectively desktop-like user experience and, above all, it enables menus and sub-windows to be rendered correctly.

Note that the X server will create a socket file inside `/tmp/.X11-unix/`. You may want to share this directory with any client of such X server, along with setting the same `DISPLAY` environment variable.

> TODO (I did not understand the previous sentence)

Reviewer: Please review this "todo"

Without noVNC, the browser has no way to show the remote display, nor to send/receive the proper VNC commands through the established websocket.

This container consists of an `nginx` server hosting the noVNC release files, in which the necessary configuration settings are dynamically injected.
This approach allows to directly use the release archive provided by the noVNC team, without the need to maintain a forked repository.

In the context of CrownLabs, this container can be used in two ways:
- as a sidecar inside the final application pod running the VNC server
- as a disjoint deployment, decoupled from the final application pod running the VNC server

In the first case, this container is tightly bounded to the VNC server, hence creating a pod that directly accepts (multiple) websocket connections.

In the second case, this container can be instantated _once_ in the entire datacenter, serving all the VNC-based containers. This results in a more efficient use of resources, but it requires the ingress controller to be able to multiplex the websocket connection towards the correct service.

> TODO (I DID NOT UNDERSTAND)

Reviewer: Please review this "todo"
